design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,282 @@
+# Thin-edge actor-based components
+
+A design goal of thin-edge is to be able to build batteries-included executables,
+with support for numerous and diverse IoT features
+that can be enabled, configured and combined by agent developers,
+to build application specific agents.
+
+For that purpose, thin-edge leverages the actor model
+and provides the tools to combine these actors into independent extensions
+and then into full-fledged executables the users can adapt to their use-cases.
+
+This document trace the core decisions toward this design.
+
+## Why actors?
+
+The core idea of the [actor model](https://en.wikipedia.org/wiki/Actor_model) is
+to implement a system with independent processing units, the actors, that
+* behave accordingly to some internal state to which they have an exclusive access,
+* send asynchronous messages to each others,
+* process, one after the other, the messages received in their inbox,
+* react to messages by sending other messages and updating their state,
+* and possibly spawn other actors.
+
+An actor can be:
+* a source of messages,
+* a mapper that consumes messages and produces translated messages,
+* a transducer that consumes and produces messages but also maintains a state ruling its behavior,
+* a server that delivers responses on requests,
+* an end-point that logs, persists or forwards the messages to some external service.
+
+In a running system, actors are connected with peers along various patterns:
+
+* An actor can gather messages from several sources and dispatch messages to several recipients.
+* Messages can be addressed to specific actors, or be broadcast to any interested recipients.
+* An actor can be oblivious to the source and the destination of the messages flowing through it,
+  processing messages independently of their source,
+  and publishing out messages to any interested recipients.
+* By contrast, an actor can distinguish its peers,
+  possibly processing the messages accordingly to their sources
+  and sending specific messages to each of them.
+* Notably, a server sends response messages specifically to the requester.
+
+The main benefits of this model in the context of thin-edge are:
+* __Modularity__
+  * An actor can be understood in isolation.
+  * The behavior of an actor is fully determined by its reaction at a given state to some given message.
+  * Actors can be implemented, tested, and packaged independently.
+* __Flexibility__
+  * Actors can be connected in various way,
+    as long as the recipients can interpret the messages received from others.
+  * An actor can be substituted by another that implements the same service,
```

Comment:
This choice of which actor to pick for the same service needs to be made when- at packaging time, declaratively or dynamically?

Reply:
Ok, fine with me.